File: src/main/java/no/nav/arbeidsgiver/min_side/services/digisyfo/SykmeldingRepository.java
package no.nav.arbeidsgiver.min_side.services.digisyfo;

import lombok.AllArgsConstructor;
import lombok.Data;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.apache.commons.lang3.tuple.Pair;

import java.util.List;
import java.util.Map;

public interface SykmeldingRepository {
    Map<String, Integer> oversiktSykmeldinger(String nærmestelederFnr);
    Map<String, Integer> oversiktSykmeldte(String nærmestelederFnr);

    @Data
    @AllArgsConstructor
    class Virksomhetsinfo {
        final String virksomhetsnummer;
        final Integer antallSykmeldte;
    }
    List<Virksomhetsinfo> sykmeldtePrVirksomhet(String nærmestelederFnr);

    void processEvent(List<ImmutablePair<String, SykmeldingHendelse>> records);
}

File: src/main/kotlin/no/nav/arbeidsgiver/min_side/services/digisyfo/SykmeldingRepository.kt
package no.nav.arbeidsgiver.min_side.services.digisyfo

import no.nav.arbeidsgiver.min_side.Database
import java.sql.ResultSet

interface SykmeldingRepository {
    suspend fun oversiktSykmeldinger(nærmestelederFnr: String): Map<String, Int>
}

class SykmeldingRepositoryImpl(
    private val database: Database
) : SykmeldingRepository {
    override suspend fun oversiktSykmeldinger(nærmestelederFnr: String): Map<String, Int> {
        return database.nonTransactionalExecuteQuery(
            """
                with nl as (
                    select virksomhetsnummer, ansatt_fnr from naermeste_leder where naermeste_leder_fnr = ?
                )
                select s.virksomhetsnummer as virksomhetsnummer, count(*) as antall
                from sykmelding as s
                join nl on
                nl.virksomhetsnummer = s.virksomhetsnummer and
                nl.ansatt_fnr = s.ansatt_fnr
                group by s.virksomhetsnummer
            """,
            { text(nærmestelederFnr) },
            { rs: ResultSet ->
                rs.getString("virksomhetsnummer") to rs.getInt("antall")
            }
        ).toMap()
    }
}


class SykmeldingRepositoryStub : SykmeldingRepository {
    override suspend fun oversiktSykmeldinger(nærmestelederFnr: String): Map<String, Int> {
        return mapOf("910825526" to 4)
    }
}
